flink/src/main/java/org/apache/iceberg/flink/source/assigner/SplitAssigner.java

```diff
@@ -0,0 +1,102 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.flink.source.assigner;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import javax.annotation.Nullable;
+import org.apache.flink.api.connector.source.SplitEnumeratorContext;
+import org.apache.iceberg.flink.source.split.IcebergSourceSplit;
+import org.apache.iceberg.flink.source.split.IcebergSourceSplitStatus;
+
+/**
+ * Enumerator should call the assigner APIs from the coordinator thread.
+ * This is to simplify the thread safety for assigner implementation.
+ */
+public interface SplitAssigner extends AutoCloseable {
+
+  /**
+   * Some assigners may need to start background threads or perform other activity such as
+   * registering as listeners to updates from other event sources e.g., watermark tracker.
+   */
+  default void start() {
+  }
+
+  /**
+   * Some assigners may need to perform certain actions
+   * when their corresponding enumerators are closed
+   */
+  @Override
+  default void close() {
+  }
+
+  /**
+   * Request a new split from the assigner
+   * as enumerator trying to assign splits to awaiting readers
+   */
+  GetSplitResult getNext(@Nullable String hostname);
+
+  /**
+   * Add new splits discovered by enumerator
+   */
+  void onDiscoveredSplits(Collection<IcebergSourceSplit> splits);
+
+  /**
+   *   Forward addSplitsBack event (for failed reader) to assigner
+   */
+  void onUnassignedSplits(Collection<IcebergSourceSplit> splits, int subtaskId);
+
+  /**
+   * Some assigner (like event time alignment) may rack in-progress splits
+   * to advance watermark upon completed splits
+   */
+  default void onCompletedSplits(Collection<String> completedSplitIds, int subtaskId) {
+  }
+
+  /**
+   * Get assigner state for checkpointing.
+   * This is a super-set API that works for all currently imagined assigners.
+   */
+  Map<IcebergSourceSplit, IcebergSourceSplitStatus> snapshotState();
+
+  /**
+   * Enumerator can get a notification via CompletableFuture
+   * when the assigner has more splits available later.
+   * Enumerator should schedule assignment in the thenAccept action of the future.
+   *
+   * Assigner will return the same future if this method is called again
+   * before the previous future is completed.
+   *
+   * The future can be completed from other thread,
+   * e.g. the coordinator thread from another thread
+   * for event time alignment.
+   *
+   * If enumerator need to trigger action upon the future completion,
+   * it may want to run it in the coordinator thread
+   * using {@link SplitEnumeratorContext#runInCoordinatorThread(Runnable)}.
+   */
+  CompletableFuture<Void> isAvailable();
+
+  /**
+   * @return assigner stats for monitoring purpose
+   */
+  SplitAssignerStats stats();
```

Comment:
I'm not sure if it makes sense to have it in SplitAssigner as this would mean every one of the implementations will need to worry about handling this logic. Apart from that, when we need to keep track of other metrics such as how long it took for a split to get assigned etc... we might need to rewrite this value class and have all assigners handle that logic. It would be better if we instead use composition here and have a listener that calculates these metrics based on certain events happening in the assigner/enumerator.

Reply:
I was using this poll model of getter on stats, because I was trying to leverage the scheduler service from enumerator to poll the stats.
```
  @Override
  public void start() {
    assigner.start();
    enumeratorContext.callAsync(
        () -> null,
        this::updateAssignerStats,
        enumeratorConfig.metricRefreshInterval().toMillis(),
        enumeratorConfig.metricRefreshInterval().toMillis());
  }
```

We can certainly go with listener model like
```
  void subscribeStats(SplitAssignerStatsListener listener);
```

But that won't make difference on those two questions raised

- each assigner implementation need to handle the stats logic
- add new stat will require change in the listener interface too, right?

Reply:
As we discussed offline, we will remove the `stats()` API from assigner. If an assigner needs to publish metrics, it should get a `MetricGroup` arg from constructor.